yt/frontends/amrvac/data_structures.py

```diff
@@ -247,81 +242,50 @@ def _set_code_unit_attributes(self):
-            else:
-                unit_length = self.units_override.get('length_unit', (1, 'cm'))
-                unit_numberdensity = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
-                unit_velocity = self.units_override.get('velocity_unit', (0, 'cm*s**-1'))
+        # Assume cgs values and let YT handle conversion if supplied in an 'mks' unit system.
+        if self.units_override:
+            unit_length = self.units_override.get('length_unit', (1, 'cm'))
+            unit_numberdensity = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
+            unit_velocity = self.units_override.get('velocity_unit', (0, 'cm*s**-1'))
             unit_temperature = self.units_override.get('temperature_unit', (1, 'K'))
-
             mylog.info('Overriding numberdensity_unit: {:1.0e} {}.'.format(unit_numberdensity[0], unit_numberdensity[1]))
-            # use only values to calculate normalisations, units are handled afterwards
-            unit_length = unit_length[0]
-            unit_numberdensity = unit_numberdensity[0]
-            unit_velocity = unit_velocity[0]
-            unit_temperature = unit_temperature[0]
         else:
             # Next, check for unit normalisations in the .dat file. Do not use .get() with default value added
             # to explicitly print a warning to the console.
             try:
-                unit_length = self.parameters['unit_length']
-                unit_numberdensity = self.parameters['unit_numberdensity']
-                unit_temperature = self.parameters['unit_temperature']
-                unit_velocity = self.parameters['unit_velocity']
-            # If not present, use default values
+                unit_length = (self.parameters['unit_length'], 'cm')
+                unit_numberdensity = (self.parameters['unit_numberdensity'], 'cm**-3')
+                unit_temperature = (self.parameters['unit_temperature'], 'K')
+                unit_velocity = (self.parameters['unit_velocity'], 'cm*s**-1')
+            # If nothing present in datfile, use default values and print warning
             except KeyError:
                 mylog.warning('No unit normalisations found in datfile and no overrides specified. Using defaults.')
-                unit_length = 1
-                unit_numberdensity = 1
-                unit_temperature = 1
-                unit_velocity = 0
+                unit_length = (1, 'cm')
+                unit_numberdensity = (1, 'cm**-3')
+                unit_temperature = (1, 'K')
+                unit_velocity = (0, 'cm*s**-1')
+
+        # Convert to YT quantities
+        unit_length = self.quan(unit_length[0], unit_length[1])
+        unit_numberdensity = self.quan(unit_numberdensity[0], unit_numberdensity[1])
+        unit_temperature = self.quan(unit_temperature[0], unit_temperature[1])
+        unit_velocity = self.quan(unit_velocity[0], unit_velocity[1])
```

Comment:
changing the type of a variable is possible in Python but should be avoided, because it's creating an additional level of overhead (you have to remember what type the variable is *according to where you are in the code*, which makes it a lot harder to work with it).
Overall, this block can and should be further reduced.